pkg/asset/installconfig/openstack/validation/platform_test.go

```diff
@@ -316,3 +319,131 @@ func TestClusterOSImage(t *testing.T) {
 		})
 	}
 }
+
+func TestMachineSubnet(t *testing.T) {
+	cases := []struct {
+		name           string
+		platform       *openstack.Platform
+		cloudInfo      *CloudInfo
+		networking     *types.Networking
+		expectedErrMsg string // NOTE: this is a REGEXP
+	}{
+		{
+			name: "external dns is not supported",
+			platform: func() *openstack.Platform {
+				p := validPlatform()
+				p.MachinesSubnet = "031a5b9d-5a89-4465-8d54-3517ec2bad48"
+				p.ExternalDNS = append(p.ExternalDNS, "1.2.3.4")
+				return p
+			}(),
+			cloudInfo:      validPlatformCloudInfo(),
+			networking:     validNetworking(),
+			expectedErrMsg: `platform.openstack.externalDNS: Invalid value: \[\]string{"1.2.3.4"}: externalDNS is set, externalDNS is not supported when machinesSubnet is set`,
+		},
+		{
+			name: "machine subnet not found",
+			platform: func() *openstack.Platform {
+				p := validPlatform()
+				p.MachinesSubnet = "031a5b9d-5a89-4465-8d54-3517ec2bad48"
+				return p
+			}(),
+			cloudInfo: func() *CloudInfo {
+				ci := validPlatformCloudInfo()
+				ci.MachinesSubnet = nil
+				return ci
+			}(),
+			networking:     validNetworking(),
+			expectedErrMsg: `platform.openstack.machinesSubnet: Not found: "031a5b9d-5a89-4465-8d54-3517ec2bad48"`,
+		},
+		{
+			name: "invalid subnet ID",
+			platform: func() *openstack.Platform {
+				p := validPlatform()
+				p.MachinesSubnet = "fake"
+				return p
+			}(),
+			cloudInfo: func() *CloudInfo {
+				ci := validPlatformCloudInfo()
+				ci.MachinesSubnet = &subnets.Subnet{
+					ID: "031a5b9d-5a89-4465-8d54-3517ec2bad48",
+				}
+				return ci
+			}(),
+			networking:     validNetworking(),
+			expectedErrMsg: `platform.openstack.machinesSubnet: Internal error: invalid subnet ID`,
+		},
+		{
+			name: "doesn't match the CIDR",
+			platform: func() *openstack.Platform {
+				p := validPlatform()
+				p.MachinesSubnet = "031a5b9d-5a89-4465-8d54-3517ec2bad48"
+				return p
+			}(),
+			cloudInfo: func() *CloudInfo {
+				ci := validPlatformCloudInfo()
+				ci.MachinesSubnet = &subnets.Subnet{
+					ID: "031a5b9d-5a89-4465-8d54-3517ec2bad48",
+				}
+				ci.MachinesSubnet.CIDR = "172.0.0.1/16"
+				return ci
+			}(),
+			networking: func() *types.Networking {
+				n := validNetworking()
+				cidr := ipnet.IPNet{
+					net.IPNet{
+						IP:   net.IPv4(172, 0, 0, 1),
+						Mask: net.IPv4Mask(255, 255, 255, 0),
+					},
+				}
+				machineNetworkEntry := &types.MachineNetworkEntry{
+					CIDR: cidr,
+				}
```

Comment:
This causes [govet to fail](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/4937/pull-ci-openshift-installer-master-govet/1405348285907996672):
```
pkg/asset/installconfig/openstack/validation/platform_test.go:405:13: github.com/openshift/installer/pkg/ipnet.IPNet composite literal uses unkeyed fields
pkg/asset/installconfig/openstack/validation/platform_test.go:436:13: github.com/openshift/installer/pkg/ipnet.IPNet composite literal uses unkeyed fields 
```

Try this:
```suggestion
				machineNetworkEntry := &types.MachineNetworkEntry{
					CIDR: *ipnet.MustParseCIDR("172.0.0.1/24"),
				}
```

Reply:
Thanks so much, it confused me for a long time. Let's wait and see:)